File: leetcode_python/Tree/number-of-nodes-in-the-sub-tree-with-the-same-label.py
"""

1519. Number of Nodes in the Sub-Tree With the Same Label
Medium

You are given a tree (i.e. a connected, undirected graph that has no cycles) consisting of n nodes numbered from 0 to n - 1 and exactly n - 1 edges. The root of the tree is the node 0, and each node of the tree has a label which is a lower-case character given in the string labels (i.e. The node with the number i has the label labels[i]).

The edges array is given on the form edges[i] = [a_i, b_i], which means there is an edge between nodes a_i and b_i in the tree.

Return an array of size n where ans[i] is the number of nodes in the subtree of the ith node which have the same label as node i.

A subtree of a tree T is the tree consisting of a node in T and all of its descendant nodes.


Example 1:

Input: n = 7, edges = [[0,1],[0,2],[1,4],[1,5],[2,3],[2,6]], labels = "abaedcd"
Output: [2,1,1,1,1,1,1]
Explanation: Node 0 has label 'a' and its sub-tree has node 2 with label 'a' as well, thus the answer is 2. Notice that any node is part of its sub-tree.
Node 1 has a label 'b'. The sub-tree of node 1 contains nodes 1,4 and 5, as nodes 4 and 5 have different labels than node 1, the answer is just 1 (the node itself).

Example 2:

Input: n = 4, edges = [[0,1],[1,2],[0,3]], labels = "bbbb"
Output: [4,2,1,1]
Explanation: The sub-tree of node 2 contains only node 2, so the answer is 1.
The sub-tree of node 3 contains only node 3, so the answer is 1.
The sub-tree of node 1 contains nodes 1 and 2, both have label 'b', thus the answer is 2.
The sub-tree of node 0 contains nodes 0, 1, 2 and 3, all with label 'b', thus the answer is 4.

Example 3:

Input: n = 5, edges = [[0,1],[0,2],[1,3],[0,4]], labels = "aabab"
Output: [3,2,1,1,1]


Constraints:

1 <= n <= 10^5
edges.length == n - 1
edges[i].length == 2
0 <= a_i, b_i < n
a_i != b_i
labels.length == n
labels is consisting of only of lowercase English letters.

"""

# V0
# IDEA : POST-ORDER DFS, MERGING 26-SLOT LABEL COUNTERS UPWARDS
#
#   for every node keep a 26-length array of label counts of its subtree.
#   process nodes in reverse DFS pre-order (= a valid post-order), so all
#   children of a node are finished before the node itself :
#
#     cnt[u] = own label + sum of cnt[child]
#     res[u] = cnt[u][labels[u]]
#
#   NOTE : n can be 10^5 and the tree may be a single 10^5-long chain, so
#          the traversal is ITERATIVE — plain recursion would overflow the
#          Python stack.
#
# time = O(26 * n), space = O(26 * n)
class Solution(object):
    def countSubTrees(self, n, edges, labels):
        g = [[] for _ in range(n)]
        for a, b in edges:
            g[a].append(b)
            g[b].append(a)

        # iterative DFS -> pre-order list + parent of each node
        parent = [-1] * n
        order = []
        seen = [False] * n
        seen[0] = True
        stack = [0]
        while stack:
            u = stack.pop()
            order.append(u)
            for v in g[u]:
                if not seen[v]:
                    seen[v] = True
                    parent[v] = u
                    stack.append(v)

        res = [0] * n
        cnt = [None] * n
        for u in reversed(order):
            cur = cnt[u]
            if cur is None:
                cur = [0] * 26
            k = ord(labels[u]) - ord('a')
            cur[k] += 1
            res[u] = cur[k]

            p = parent[u]
            if p >= 0:
                if cnt[p] is None:
                    cnt[p] = [0] * 26
                pc = cnt[p]
                for i in range(26):
                    pc[i] += cur[i]
            cnt[u] = None   # free it, the parent already absorbed it
        return res

File: leetcode_java/src/main/java/LeetCodeJava/Tree/NumberOfNodesInTheSubTreeWithTheSameLabel.java
package LeetCodeJava.Tree;

// https://leetcode.com/problems/number-of-nodes-in-the-sub-tree-with-the-same-label/

/**
 *  1519. Number of Nodes in the Sub-Tree With the Same Label
 *  Medium
 *
 *  You are given a tree (i.e. a connected, undirected graph that has no
 *  cycles) consisting of n nodes numbered from 0 to n - 1 and exactly n - 1
 *  edges. The root of the tree is node 0, and each node has a label which is a
 *  lower-case character given in the string labels (the node numbered i has
 *  the label labels[i]).
 *
 *  The edges array is given on the form edges[i] = [a_i, b_i], meaning there
 *  is an edge between nodes a_i and b_i.
 *
 *  Return an array of size n where ans[i] is the number of nodes in the
 *  subtree of the ith node which have the same label as node i.
 *
 *  Example 1:
 *    Input: n = 7, edges = [[0,1],[0,2],[1,4],[1,5],[2,3],[2,6]],
 *           labels = "abaedcd"
 *    Output: [2,1,1,1,1,1,1]
 *    Explanation: node 0 has label 'a' and its subtree also holds node 2
 *                 with label 'a', so the answer is 2.
 *
 *  Example 2:
 *    Input: n = 4, edges = [[0,1],[1,2],[0,3]], labels = "bbbb"
 *    Output: [4,2,1,1]
 *
 *  Constraints:
 *    1 <= n <= 10^5
 *    edges.length == n - 1
 *    0 <= a_i, b_i < n, a_i != b_i
 *    labels.length == n, lowercase English letters only
 */
public class NumberOfNodesInTheSubTreeWithTheSameLabel {

    // V0
    // IDEA: POST-ORDER OVER AN ITERATIVE DFS ORDER, MERGING 26-SLOT COUNTERS UP
    //       for every node keep a 26-length array of the label counts of its
    //       subtree. walking the DFS pre-order list in REVERSE is a valid
    //       post-order, so all children of a node are finished before it:
    //           cnt[u] = own label + sum of cnt[child]
    //           res[u] = cnt[u][labels[u]]
    //       NOTE: n can be 10^5 and the tree may be one long chain, so the
    //             traversal is iterative (recursion could blow the stack).
    /**
     * time = O(26 * N)
     * space = O(26 * N)
     */
    public int[] countSubTrees(int n, int[][] edges, String labels) {
        // adjacency via head/next arrays (no boxing)
        int[] head = new int[n];
        int[] nxt = new int[2 * (n - 1 > 0 ? n - 1 : 1)];
        int[] to = new int[2 * (n - 1 > 0 ? n - 1 : 1)];
        for (int i = 0; i < n; i++) {
            head[i] = -1;
        }
        int ec = 0;
        for (int[] e : edges) {
            to[ec] = e[1]; nxt[ec] = head[e[0]]; head[e[0]] = ec++;
            to[ec] = e[0]; nxt[ec] = head[e[1]]; head[e[1]] = ec++;
        }

        // iterative DFS -> pre-order list + parent of each node
        int[] parent = new int[n];
        int[] order = new int[n];
        boolean[] seen = new boolean[n];
        int[] stack = new int[n];
        int sp = 0, oc = 0;
        parent[0] = -1;
        seen[0] = true;
        stack[sp++] = 0;
        while (sp > 0) {
            int u = stack[--sp];
            order[oc++] = u;
            for (int e = head[u]; e != -1; e = nxt[e]) {
                int v = to[e];
                if (!seen[v]) {
                    seen[v] = true;
                    parent[v] = u;
                    stack[sp++] = v;
                }
            }
        }

        int[] res = new int[n];
        int[][] cnt = new int[n][];
        for (int i = oc - 1; i >= 0; i--) {
            int u = order[i];
            int[] cur = cnt[u];
            if (cur == null) {
                cur = new int[26];
            }
            int k = labels.charAt(u) - 'a';
            cur[k]++;
            res[u] = cur[k];

            int p = parent[u];
            if (p >= 0) {
                if (cnt[p] == null) {
                    cnt[p] = new int[26];
                }
                int[] pc = cnt[p];
                for (int j = 0; j < 26; j++) {
                    pc[j] += cur[j];
                }
            }
            cnt[u] = null;   // parent already absorbed it -> free it
        }
        return res;
    }
}
